pya/aserver.py

```diff
@@ -12,6 +13,126 @@
+        return self._signal
+
+    @property
+    def remaining_samples(self) -> int:
+        return self._signal.shape[0] - self.pos
+
+    def get_next_chunk(self, max_samples: int) -> np.ndarray:
+        """Retrieves the next portion of audio samples from the signal, advancing the playback position.
+        The function handles both partial and complete chunks, automatically marking the event as completed
+        when all samples have been played.
+
+        Parameters
+        ----------
+        max_samples : int
+            Maximum number of samples to return
+
+        Returns
+        -------
+        np.ndarray
+            Array of audio samples, empty if event is completed
+        """
+        if self.completed:
+            return np.array([])
+
+        samples = min(max_samples, self.remaining_samples)
+        chunk = self._signal[self._pos: self._pos + samples]
+        self._pos += samples
+
+        if self._pos >= self._signal.shape[0]:
+            self.completed = True
+
+        return chunk
+
+
+class EventScheduler:
+    """Manages scheduling of audio events for playback."""
+
+    def __init__(self, sr: int):
+        self._sr = sr
+        self.pending_events = []
+        self.active_events = []
+
+    def schedule(self, event: AudioEvent):
+        idx = 0
+        while (idx < len(self.pending_events) and self.pending_events[idx].onset < event.onset):
+            idx += 1
+
+        self.pending_events.insert(idx, event)
+
+    def process(
+        self, current_time: float, output_buffer: np.ndarray, block_size: int
+    ) -> np.ndarray:
+        end_time = current_time + block_size / self._sr
+
+        idx = 0
+        while idx < len(self.pending_events):
+            if self.pending_events[idx].onset <= end_time:
+                self.active_events.append(self.pending_events.pop(idx))
+            else:
+                idx += 1
+
+        idx = 0
+        while idx < len(self.active_events):
+            event = self.active_events[idx]
+
+            if event.onset > current_time:
+                offset_samples = int((event.onset - current_time) * self._sr)
+            else:
+                offset_samples = 0
+
+            audio_chunk = event.get_next_chunk(block_size - offset_samples)
+
+            if len(audio_chunk) > 0:
+                channels = min(
+                    audio_chunk.shape[1] if len(audio_chunk.shape) > 1 else 1,
+                    output_buffer.shape[1] - event.out_channel,
+                )
+
+                if len(audio_chunk.shape) == 1:
+                    audio_chunk = audio_chunk.reshape(-1, 1)
+
+                end_offset = offset_samples + audio_chunk.shape[0]
+                end_channel = event.out_channel + channels
+
+                output_buffer[
+                    offset_samples:end_offset, event.out_channel: end_channel
+                ] += audio_chunk[:, :channels]
+
+            if event.completed:
+                self.active_events.pop(idx)
```

Comment:
This results potentially in many pop events. Instead I had (in my original code) a `dellist` collecting indices of completed entries and performed deletions on all completed items after processing all playing. I can't remember my analysis (years ago) how much that reduced time to have the list maintenance 'in one go', so its difficult to judge the effect of your rewriting, but I'd guess that this might be a bit slower. Anyway, I wonder whether `del active_events[idx]` is faster than `.pop(idx)` and OK as the result of `.pop(idx)` is not needed any further...